### ISSUE TITLE
Azure Service Bus SessionReceiver Retry TryTimeout Behavior

### ISSUE THREAD

**Issue by JRiggenbach** [Azure/azure-sdk-for-java#32455]
Did the amqp retry behavior change since this post?
https://github.com/Azure/azure-sdk-for-java/issues/18778

`.retryOptions(new AmqpRetryOptions().setTryTimeout(Duration.ofSeconds(20)))`

For the life of me, I can't get the processor client to grab onto new sessions any faster than a 60 second interval.  

Here is my ServiceBusClientBuilder config

```
return new ServiceBusClientBuilder()
    .credential(serviceBusNamespace, clientSecretCredential)
    .retryOptions(new AmqpRetryOptions()
        .setTryTimeout(Duration.ofSeconds(5)))
    .sessionProcessor()
    .topicName(topicName)
    .subscriptionName(subscriptionName)
    .receiveMode(ServiceBusReceiveMode.PEEK_LOCK)
    .maxConcurrentSessions(10)
    .processMessage(messageHandler::processMessage)
    .processError(messageHandler::processError)
    .buildProcessorClient();
```

I would expect this to generate a client that would receive message from 10 session (which it does) and if there are no messages received in 5 seconds for a session, it would move on to next session and start getting message from next available session.  However, the next sessions are not processed until that 60 second loop is complete.  I can see that the duration is applied to the session and that it reports out that there are no other messages within that 5 second window, but I must be missing something in my understanding of how this should work.

FWIW, I do see quite a few "Error occurred while getting unnamed session." log entries

> {"timestamp":"2022-12-05 19:41:49.260","message":"{\"az.sdk.message\":\"Error occurred while getting unnamed session.\",\"exception\":\"The operation did not complete within the allotted timeout of 00:00:04. The time allotted to this operation may have been a portion of a longer timeout. For more information on exception types and proper exception handling, please refer to https://aka.ms/ServiceBusExceptions . TrackingId:47be4424-fdd8-4481-bc7e-1cbfd628036a_B11, SystemTracker:bach-cps-vendor-comm-service-bus:Topic:eligibility|autoresolverefill, Timestamp:2022-12-06T00:41:49 TrackingId:68aaed80ab4f4e75904c619d080f2887_G39, SystemTracker:gateway7, Timestamp:2022-12-06T00:41:49, errorContext[NAMESPACE: bach-cps-vendor-comm-service-bus.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: eligibility/subscriptions/AutoResolveRefill, REFERENCE_ID: session-_4e373c_1670287305150, LINK_CREDIT: 0]\",\"entityPath\":\"eligibility/subscriptions/AutoResolveRefill\",\"attempt\":5}

I can trace that exception back to The [ServiceBusSessionManager ](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSessionManager.java#L274)and from there it appears that there is a hard-coded 1 minute delay that can't be changed.  

Additionally, if I start my application and there are no messages on the subscription, when I do publish messages to the topic, it takes however long I'm into the 60 second loop to complete before the messages begin to be processed.   I found this [post ](https://github.com/Azure/azure-service-bus-java/issues/394) so it appears that the old SDK did the same thing at one time but was addressed.  Now the new SDK has the same issue?

I'm using 
```
<dependency>
    <groupId>com.azure</groupId>
    <artifactId>azure-messaging-servicebus</artifactId>
    <version>7.13.0</version>
</dependency>
```

Java 11 with Spring Boot 2.5.5.  

Thanks.

_Originally posted by @JRiggenbach in https://github.com/Azure/azure-sdk-for-java/issues/18778#issuecomment-1338291995_

**Comment by joshfree** [comment]
@ZejiaJiang could you please follow up with @JRiggenbach on this github issue?

**Comment by ZejiaJiang** [comment]
Hi @JRiggenbach ,

Thank you for your report.

 I just go through your case and the original issue. I'm trying to repro it.

**Comment by ZejiaJiang** [comment]
Hi @JRiggenbach ,

I didn't meet your case. Hence I'd like to clarify your scenario, do you mean that you have 10 sessions and need to process message from these sessions? But it will be timeout because there is a long time between 2 messages received?

if you have 10 sessions in total, you can use the builder code that you pasted and set a large duration for timeout, because you've already listen 10 sessions concurrently, no need to set a small timeout to retry.

Actually I don't think you need session feature, it seems like that there is a long time between two messages received, just use the processor client is enough. Session feature is for guaranteeing ordered delivery. So would you mind to talk more about your scenario, I'll help you to use this SDK.

**Comment by JRiggenbach** [comment]
@ZejiaJiang 
Thanks for looking to this question with me.  We do need sessions; within our production environments we deal with millions of transactions and come all throughout the day.  This was just something we experienced when cutting over to the new SDK and wanted to understand if there was a way to work around it.

As I stated in the original post, when I start my application and then publish a message to my session aware topic/queue, the message will not be picked up until the current 60 second wait period is over.

**Comment by ZejiaJiang** [comment]
hi @JRiggenbach ,

Thank you for you explanation. 

Could you please provide your code snippet for sending and receiving messages? It seems like that there are something wrong in your configuration or your code. Let me help you to receive message first.

 I ran sample code and it works well. you can also run the sample code with very default configuration to test it works, and then change your configuration.

**Comment by JRiggenbach** [comment]
I opened a support ticket via my Azure subscription support plan.  The tracking id for that ticket is #2212190040006676.  I had to go through the same initial conversations with that ticket as am I here, but Microsoft finally acknowledged that this is an issue.  On the 9th of January, I was informed that the product team will be mitigating the issue by Feb 2nd week at the earliest, but I have yet to see an issue to track the progress.

**Comment by JRiggenbach** [comment]
The issue can easily be replicated using the following code
https://github.com/JRiggenbach/azure-service-bus-session-sample

1. Start the app
2. Publish a message to the service bus

Result - will take up to 60 seconds before the message is processed.

**Comment by ZejiaJiang** [comment]
Hi @JRiggenbach , thank you for your code and sorry for delay reply.

I ignored there is no messages in topic before, so I didn't repro it. I've met your issue this time. 

here is a quick workround, set `maxConcurrentSessions` as how many sessions you have. don't set `setTryTimeout` or set a large duration such as 2mins.

the root cause is when there is no message received from link and we set `tryTimeout` 5 sec. create link method will throw a timeout error and retry after 1min in case of networking is bad or something wrong and client retry too many times.

I'll dive deep into the code and discuss with coworker what's best practice in this case. 

overall, the qiuck workaround is I mentioned above or sent some messages to the topic before the processor client start.

**Comment by JRiggenbach** [comment]
@ki1729 Thanks for the change.  Unfortunately we're still not seeing the behavior we think we should be seeing, but it looks like we'll be talking to each other on Thursday.  Hopefully there is just some misunderstanding/misconfiguration on our end.

**Comment by ki1729** [comment]
Ah yes we are! Looking forward to it, hopefully we can figure out a resolution.

**Comment by gaderGarcia** [comment]
what was the outcome of this result, is there any version available to mitigate this issue?

**Comment by JRiggenbach** [comment]
Yes, the issue is addressed in the 7.13.4 release
